lib/charms/operator_libs_linux/v0/sysctl.py

```diff
@@ -27,6 +87,211 @@
+            f.writelines(charm_params)
+
+    def _merge(self, add_own_charm=True) -> None:
+        """Create the merged sysctl file.
+
+        Args:
+            add_own_charm : bool, if false it will skip the charm file from the merge.
+        """
+        # get all files that start by 90-juju-
+        data = [SYSCTL_HEADER]
+        paths = set(SYSCTL_DIRECTORY.glob(f"{CHARM_FILENAME_PREFIX}*"))
+        if not add_own_charm:
+            paths.discard(self.charm_filepath.as_posix())
+
+        for path in paths:
+            with open(path, "r") as f:
+                data += f.readlines()
+        with open(SYSCTL_FILENAME, "w") as f:
+            f.writelines(data)
+
+        # Reload data with newly created file.
+        self._data = self._load_data()
+
+    def _apply(self) -> None:
+        """Apply values to machine."""
+        cmd = [f"{key}={value}" for key, value in self._desired_config.items()]
+        result = self._sysctl(cmd)
+        expr = re.compile(r"^sysctl: permission denied on key \"([a-z_\.]+)\", ignoring$")
+        failed_values = [expr.match(line) for line in result if expr.match(line)]
+        logger.debug("Failed values: %s", failed_values)
+
+        if failed_values:
+            msg = f"Unable to set params: {[f.group(1) for f in failed_values]}"
+            logger.error(msg)
+            raise SysctlPermissionError(msg)
+
+    def _create_snapshot(self) -> Dict[str, str]:
+        """Create a snaphot of config options that are going to be set."""
+        return {key: self._sysctl([key, "-n"])[0] for key in self._desired_config.keys()}
+
+    def _restore_snapshot(self, snapshot: Dict[str, str]) -> None:
+        """Restore a snapshot to the machine."""
+        values = [f"{key}={value}" for key, value in snapshot.items()]
+        self._sysctl(values)
+
+    def _sysctl(self, cmd: List[str]) -> List[str]:
+        """Execute a sysctl command."""
+        cmd = ["sysctl"] + cmd
+        logger.debug("Executing sysctl command: %s", cmd)
+        try:
+            return check_output(cmd, stderr=STDOUT, universal_newlines=True).splitlines()
+        except CalledProcessError as e:
+            msg = f"Error executing '{cmd}': {e.stdout}"
+            logger.error(msg)
+            raise SysctlError(msg)
+
+    def _parse_config(self, config: Dict[str, dict]) -> None:
+        """Parse a config passed to the lib."""
+        result = {}
+        for key, value in config.items():
+            result[key] = str(value["value"])
+        self._desired_config: Dict[str, str] = result
+
+    def _load_data(self) -> Dict[str, str]:
+        """Get merged config."""
+        config = {}
+        if not SYSCTL_FILENAME.exists():
+            return config
+
+        with open(SYSCTL_FILENAME, "r") as f:
+            for line in f:
+                config.update(self._parse_line(line))
+
+        return config
+
+    def _parse_line(self, line: str) -> Dict[str, str]:
```

Comment:
To me it would be much cleaner if this function returns `Tuple[str, str]` and not dictionary with single key and value.